### PR TITLE
IBX-810: Extendable repository configuration

### PR DESCRIPTION
**Pull request by adamwojs**
| Question                                  | Answer
| ---------------------------------------- | ------------------
| **JIRA issue**                          | https://issues.ibexa.co/browse/IBX-810
| **Type**                                   | feature
| **Target eZ Platform version** | `v4.0`
| **BC breaks**                          | no
| **Doc needed**                       | yes

Added extension point which allows to define custom semantic configuration under `ezpublish.repositories.<name>` key.

### Usage

1. Implement `\eZ\Bundle\EzPublishCoreBundle\DependencyInjection\Configuration\RepositoryConfigParserInterface`

```php
use Symfony\Component\Config\Definition\Builder\NodeBuilder;

interface RepositoryConfigParserInterface
{
    public function addSemanticConfig(NodeBuilder $nodeBuilder): void;
}
```

e.g.

```php
use eZ\Bundle\EzPublishCoreBundle\DependencyInjection\Configuration\RepositoryConfigParserInterface;
use Symfony\Component\Config\Definition\Builder\NodeBuilder;

final class CustomRepositoryConfigParser implements RepositoryConfigParserInterface
{
    public function addSemanticConfig(NodeBuilder $nodeBuilder): void
    {
        $nodeBuilder
            ->arrayNode('acme')
                ->children()
                    ->scalarNode('coyote_speed')
                        ->info('Maximum coyote speed (km/h) ')
                        ->isRequired()
                        ->defaultValue(120)
                    ->end()
                    // ...
                ->end()
            ->end();
    }
}
```

More examples can be found under `eZ\Bundle\EzPublishCoreBundle\DependencyInjection\Configuration\Parser\Repository` namespace.

2. Register repository configuration extension:

```php
final clas AcmeFeatureBundle extends Bundle
{
    public function build(ContainerBuilder $container): void
    {
        // ...

        /** @var \eZ\Bundle\EzPublishCoreBundle\DependencyInjection\EzPublishCoreExtension $kernel */
        $kernel = $container->getExtension('ezpublish');
        $kernel->addRepositoryConfigParser(new CustomRepositoryConfigParser());
    }
}
```

3. Access configuration using `\eZ\Bundle\EzPublishCoreBundle\ApiLoader\RepositoryConfigurationProvider::getRepositoryConfig` method

```php
$acmeConfig = $repositoryConfigProvider->getRepositoryConfig()['acme'];
```

#### Checklist:
- [x] Provided PR description.
- [x] Tested the solution manually.
- [x] Provided automated test coverage.
- [x] Checked that target branch is set correctly (master for features, the oldest supported for bugs).
- [x] Ran PHP CS Fixer for new PHP code (use `$ composer fix-cs`).
- [x] Asked for a review (ping `@ezsystems/php-dev-team`).

### CODE REVIEW

File: eZ/Bundle/EzPublishCoreBundle/DependencyInjection/EzPublishCoreExtension.php [hunk 1]
```diff
@@ -181,18 +195,33 @@ public function prepend(ContainerBuilder $container)
...
-            $this->mainConfigParser = new ConfigParser($this->configParsers);
+            $this->mainConfigParser = new ConfigParser($this->siteAccessConfigParsers);
         }
 
         return $this->mainConfigParser;
     }
 
+    private function getMainRepositoryConfigParser(): RepositoryConfigParserInterface
+    {
+        if ($this->mainRepositoryConfigParser === null) {
```
Comment (Steveb-p) [inline eZ/Bundle/EzPublishCoreBundle/DependencyInjection/EzPublishCoreExtension.php:212]: To allow non-nullable typed property to work once the code gets updated to PHP 7.4+.
```suggestion
        if (!isset($this->mainRepositoryConfigParser)) {
```

File: src/bundle/Core/DependencyInjection/Configuration/RepositoryConfigParser.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+final class RepositoryConfigParser implements RepositoryConfigParserInterface
+{
+    /** @var \Ibexa\Bundle\Core\DependencyInjection\Configuration\RepositoryConfigParserInterface[] */
+    private $configParsers;
+
+    /**
+     * @param \Ibexa\Bundle\Core\DependencyInjection\Configuration\RepositoryConfigParserInterface[] $configParsers
+     */
+    public function __construct(array $configParsers = [])
```
Comment (Steveb-p) [inline src/bundle/Core/DependencyInjection/Configuration/RepositoryConfigParser.php:22]: Optional: Argument and property can be made `iterable`. We're not using functionalities specific to arrays anywhere.

**Review by Steveb-p — APPROVED** [review]

**Review by mikadamczyk — APPROVED** [review]

File: src/bundle/Core/DependencyInjection/Configuration/RepositoryConfigParser.php (hunk 2 above)
Comment (adamwojs) [reply src/bundle/Core/DependencyInjection/Configuration/RepositoryConfigParser.php]: Fixed in https://github.com/ezsystems/ezplatform-kernel/pull/218/commits/69ff21ef0a40cbf323d7ef379f1bcaaffdca4f64

File: eZ/Bundle/EzPublishCoreBundle/DependencyInjection/EzPublishCoreExtension.php (hunk 1 above)
Comment (adamwojs) [reply eZ/Bundle/EzPublishCoreBundle/DependencyInjection/EzPublishCoreExtension.php]: Fixed in https://github.com/ezsystems/ezplatform-kernel/pull/218/commits/69ff21ef0a40cbf323d7ef379f1bcaaffdca4f64

**Review by mnocon — APPROVED** [review]
Tested with my own Command, thanks for great PR description how to set it up!

I was able to create my own setting, define it's value for different repositories (which I configured for different SiteAccesses) and verify with the Command that correct value is used for each SA. Also tested some existing settings, also went ok.

It also behaved ok when the repo was defined on SiteAccess group level.

**One nitpick for the doc:**
- change `eZ\Bundle\EzPublishCoreBundle\DependencyInjection\Configuration\RepositoryConfigParserInterface` to `Ibexa\Bundle\Core\DependencyInjection\Configuration\RepositoryConfigParserInterface`, I guess it was rebranded after the description has been created.

**Pull request merged**